internal/tools/runner.go

```diff
@@ -0,0 +1,308 @@
+				env = append(env, key+"="+val)
+				passedKeys = append(passedKeys, key)
+			}
+		}
+	}
+	return env, passedKeys
+}
+
+// normalizeWaitError maps timeout and process errors to deterministic errors.
+func normalizeWaitError(ctx context.Context, waitErr error, stderrText string) error {
+	if ctx.Err() == context.DeadlineExceeded {
+		return errors.New("tool timed out")
+	}
+	if waitErr != nil {
+		msg := stderrText
+		if msg == "" {
+			msg = waitErr.Error()
+		}
+		return errors.New(msg)
+	}
+	return nil
+}
+
+func RunToolWithJSON(parentCtx context.Context, spec ToolSpec, jsonInput []byte, defaultTimeout time.Duration) ([]byte, error) {
+	start := time.Now()
+	// Derive timeout, honoring per-tool override when provided.
+	to := computeToolTimeout(spec, defaultTimeout)
+	ctx, cancel := context.WithTimeout(parentCtx, to)
+	defer cancel()
+
+	cmd := exec.CommandContext(ctx, spec.Command[0], spec.Command[1:]...)
+	// Build minimal environment and record passed-through keys for audit.
+	env, passedKeys := buildToolEnvironment(spec)
+	cmd.Env = env
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, fmt.Errorf("stdin pipe: %w", err)
+	}
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, fmt.Errorf("stdout pipe: %w", err)
+	}
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, fmt.Errorf("stderr pipe: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, fmt.Errorf("start: %w", err)
+	}
+	// Write JSON to stdin
+	if len(jsonInput) == 0 {
+		jsonInput = []byte("{}")
+	}
+	if _, err := stdin.Write(jsonInput); err != nil {
+		return nil, fmt.Errorf("write stdin: %w", err)
+	}
+	// Best-effort close; log failure to audit but do not fail run
+	if err := stdin.Close(); err != nil {
+		// Capture the close error as a best-effort audit line
+		if err2 := appendAuditLog(map[string]any{
+			"ts":    timeNow().UTC().Format(time.RFC3339Nano),
+			"event": "stdin_close_error",
+			"tool":  spec.Name,
+			"error": err.Error(),
+		}); err2 != nil {
+			_ = err2
+		}
+	}
+
+	// Read stdout and stderr fully
+	outCh := make(chan []byte, 1)
+	errCh := make(chan []byte, 1)
+	go func() { outCh <- safeReadAll(stdout) }()
+	go func() { errCh <- safeReadAll(stderr) }()
+
+	err = cmd.Wait()
+	out := <-outCh
+	serr := <-errCh
+
+	exitCode := 0
+	if err != nil {
+		// Try to capture exit code when available
+		if ee, ok := err.(*exec.ExitError); ok && ee.ProcessState != nil {
+			exitCode = ee.ProcessState.ExitCode()
+		} else {
+			// Unknown exit (e.g., timeout/cancel)
+			exitCode = -1
+		}
+	}
+	// Best-effort audit (failures do not affect tool result)
+	writeAudit(spec, start, exitCode, len(out), len(serr), passedKeys)
+
+	if normErr := normalizeWaitError(ctx, err, string(serr)); normErr != nil {
+		return nil, normErr
+	}
+	return out, nil
+}
+
+// safeReadAll reads all bytes from r; on error it returns any bytes read so far (or nil).
+func safeReadAll(r io.Reader) []byte {
```

Comment:
I think this could be moved to some helper file. We should have all functions grouped in smaller semantic segments in their own files, or even their own modules.